### PR TITLE
feat: support creating endpoints with model images from private registries

### PR DESCRIPTION
**Pull request by wuyrush**
*Issue #, if available:*

*Description of changes:*
Allow ImageConfig to be passed as a parameter when creating endpoints.

Reference:
* https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_ImageConfig.html
* https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sagemaker.html#SageMaker.Client.create_model

*Testing done:*
- Unit tests
- Create SageMaker endpoint using model image from private registry with SageMaker Python SDK and verified creation succeeded

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/image_config.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                * Platform: The model image is hosted in Amazon ECR.
+                * Vpc: The model image is hosted in a private Docker registry in your VPC.
+        """
+        self.repository_access_mode = repository_access_mode
+
+    def _to_request_dict(self):
+        """Generates a request dictionary using the parameters provided to the class."""
+        req = {
+            "RepositoryAccessMode": "Platform",
+        }
```
Comment (metrizable) [inline src/sagemaker/image_config.py:37]: this generates a new `dict` every time the method is invoked. is that desired?

File: src/sagemaker/model.py [hunk 2]
```diff
@@ -127,7 +133,13 @@ def prepare_container_def(
         Returns:
             dict: A container definition object usable with the CreateModel API.
         """
-        return sagemaker.container_def(self.image_uri, self.model_data, self.env)
+        image_config_dict = None
+        if self.image_config:
+            image_config_dict = self.image_config._to_request_dict()
```
Comment (metrizable) [inline src/sagemaker/model.py:138]: the encapsulating class `ImageConfig` leads to extra code

File: src/sagemaker/image_config.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""This module contains code to create and manage SageMaker ``ImageConfig``"""
+from __future__ import absolute_import
+
+
+class ImageConfig(object):
+    """Configuration of Docker image used in Model."""
```
[the remark this answers is not in the document]
Comment (wuyrush) [reply src/sagemaker/image_config.py:18]: > is this class needed? it's not clear what value it is providing. it's essentially a dict with a single key and default value.

The initial idea of using such class is to retain enough flexibility as more features for private registry support are added in the future. Thinking twice, this seems an overkill as `ImageConfig` has already positioned itself as a configuration bag, so I think a plain `dict[str, object]` shall suffice.

**Review by icywang86rui — APPROVED** [review]

**Comment by icywang86rui** [comment]
please resolve the conflict

**Comment by icywang86rui** [comment]
Please resolve the conflicts.

**Review by ChoiByungWook — APPROVED** [review]
Awesome!!!

I resolved the merge conflict and also submitted a few commits to address pylint issues.

**Pull request merged**